elephant/joint_isi_dithering_class.py

```diff
@@ -0,0 +1,462 @@
+        Wutside of it is only necessary if the attributes of the
+        :class:`Joint_ISI_Space` were changed after the initialization, than it
+        prepares the class again to create dithered spiketrains.
+    dithering()
+        Returns a list of dithered spiketrains and if print_mode it returns
+        also a string 'uniform' or 'jisid' indicating the way, how the dithered
+        spiketrains were obtained.
+
+    Output:
+    dithered_sts
+        List of spiketrains, dithered with Joint-ISI-dithering
+    mode (only if print_mode=True)
+        string: Indicates, which method was used to dither the spikes.
+            'jisih' if joint-ISI was used,
+            'uniform' if the dense_rate was too low and uniform dithering was
+             used.
+    '''
+
+    def __init__(self,
+                 st,
+                 n_surr=1,
+                 dither=15.*pq.ms,
+                 unit=pq.s,
+                 window_length=120.*pq.ms,
+                 num_bins=120,
+                 sigma=1.*pq.ms,
+                 isi_median_threshold=30*pq.ms,
+                 alternate=True,
+                 show_plot=False,
+                 print_mode=False,
+                 use_sqrt=False,
+                 method='fast',
+                 cutoff=True,
+                 min_spikes=10
+                 ):
+        self.st = st
+
+        self.n_surr = n_surr
+
+        self.unit = unit
+        self.dither = dither
+        self.window_length = window_length
+        self.sigma = sigma
+        self.isi_median_threshold = isi_median_threshold
+
+        self.num_bins = num_bins
+
+        self.alternate = alternate
+        self.show_plot = show_plot
+        self.print_mode = print_mode
+        self.use_sqrt = use_sqrt
+        self.method = method
+        self.cutoff = cutoff
+        self.min_spikes = min_spikes
+
+        self.preprocessing()
+
+    def preprocessing(self):
+        '''
+        All preprocessing steps for the joint-ISI dithering are done here.
+
+        So first to checks are done. If they are not passed, self.method is
+        set to 'uniform'. The first one asks for the number of spikes.
+        The second compares the median of the ISI-distribution against a
+        threshold.
+
+        If the method is not 'uniform' the cumulative distribution functions
+        for the Joint-ISI dither process are evaluated.
+
+        If method is 'fast':
+        For each slice of the joint-ISI
+        distribution (parallel to the antidiagonal) a cumulative distr.
+        function is calculated.
+
+        If method is 'window':
+        For each point in the joint-ISI distribution a on the line parallel to
+        the antidiagonal all points up to the dither-parameter are included, to
+        calculate the cumulative distribution function.
+
+        The function has no output, but stores its result inside the class.
+        '''
+        if len(self.st) < self.min_spikes:
```

Comment:
I would shorten the explanation in the documentation, but add it here in the code, check by check, step by step